README.md
# tcping

[![npm](https://img.shields.io/npm/v/@discretetom/tcping?color=green&style=flat-square)](https://www.npmjs.com/package/@discretetom/tcping)
![license](https://img.shields.io/github/license/DiscreteTom/tcping?style=flat-square)

A CLI tool to test how long it takes to establish an TCP connection to a specific address, with percentile support (e.g. p99).

## Installation

```sh
npm install -g @discretetom/tcping
```

## Usage

```
# tcping --help
Usage: tcping [options] <host> [port]
Usage (HTTP mode): tcping -h [options] <url>

Options:
      --help        Show help                                          [boolean]
      --version     Show version number                                [boolean]
  -i, --interval                                         [number] [default: 500]
  -t, --timeout                                         [number] [default: 3000]
  -p, --port        Port number. Not effective in HTTP mode.
                                                          [number] [default: 80]
  -c, --count       Number of requests to send. -1 for infinite.
                                                          [number] [default: -1]
  -P, --percentile  A list of percentile to display, separated by ",", e.g.
                    "95,99".                              [string] [default: ""]
  -T, --timestamp   Display timestamp in output.      [boolean] [default: false]
  -h, --http        Use HTTP mode.                    [boolean] [default: false]
```

## Examples

### Basics

```
# tcping baidu.com 443
host: baidu.com, port: 443, count: Infinity
finish dns lookup, address: 39.156.66.10 family: IPv4

connected: 52.546ms
connected: 54.906ms
connected: 50.258ms
connected: 50.999ms
connected: 56.020ms
connected: 54.988ms
connected: 344.278ms
connected: 46.484ms

statistics:
sent: 8, received: 8, success rate: 100%
min: 46.484ms, max: 344.278ms, avg: 88.810ms
```

### Percentile

```
# tcping baidu.com 80 -P 95
host: baidu.com, port: 80, count: Infinity
finish dns lookup, address: 39.156.66.10 family: IPv4

connected: 10.422ms
connected: 7.141ms
connected: 8.139ms
connected: 8.767ms
connected: 10.574ms
connected: 8.519ms
connected: 8.372ms
connected: 7.644ms
connected: 8.428ms
connected: 8.558ms
connected: 18.181ms
connected: 9.103ms

statistics:
sent: 12, received: 12, success rate: 100.00%
min: 7.141ms, max: 18.181ms, avg: 9.487ms, p95: 18.181ms
```

### HTTP Mode

```
# tcping -h https://baidu.com
HTTP mode, url: https://baidu.com, count: Infinity

get: 100.288ms, status: 302, length: 161
get: 25.267ms, status: 302, length: 161
get: 26.985ms, status: 302, length: 161
get: 24.540ms, status: 302, length: 161
get: 24.518ms, status: 302, length: 161

statistics:
sent: 5, received: 5, success rate: 100.00%
min: 24.518ms, max: 100.288ms, avg: 40.320ms
```

## [CHANGELOG](https://github.com/DiscreteTom/tcping/blob/main/CHANGELOG.md)


CHANGELOG.md
# CHANGELOG

## v0.1.4

- Add HTTP mode.

## v0.1.3

- Optimize performance.
- Optimize code.

## v0.1.2

- Fix success rate value and format.
- Fix: only update `stats.sent` on success / failure.
- New option: `--timestamp` / `-T`.

## v0.1.1

- New option: `--count` / `-c`.
- New option: `--percentile` / `-P`.
- Handle Ctrl-C.
- Print stats.

## v0.1.0

Initial release.


src/lib.ts
import net from "net";
import dns from "dns";
import { config } from "./config";
import { stats } from "./stats";
import * as https from "https";
import * as http from "http";

async function sleep(ms: number) {
  return new Promise<void>((resolve) => setTimeout(resolve, ms));
}

/** Get current time in microseconds. */
function now() {
  const hrtime = process.hrtime();
  return hrtime[0] * 1000000 + hrtime[1] / 1000;
}

export async function run() {
  console.log(
    `host: ${config.host}, port: ${config.port}, count: ${config.count}`
  );

  // check whether host is ip address using regex
  if (!/^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}$/.test(config.host)) {
    // not ip address, dns lookup
    try {
      await new Promise<void>((resolve, reject) => {
        dns.lookup(config.host, (err, address, family) => {
          if (err) {
            reject(err);
            return;
          }
          console.log(
            `finish dns lookup, address: ${address} family: IPv${family}`
          );
          resolve();
        });
      });
    } catch (e) {
      console.error(e);
      return;
    }
  }

  console.log(); // empty line

  while (stats.sent < config.count) {
    try {
      await new Promise<void>((resolve, reject) => {
        const sock = new net.Socket();
        sock.setTimeout(config.timeout);

        const start = now();
        sock
          .on("connect", function () {
            const latencyMs = (now() - start) / 1000;
            const result =
              (config.timestamp ? new Date().toISOString() + " " : "") +
              `connected: ${latencyMs.toFixed(3)}ms`;
            console.log(result);
            sock.destroy();

            // update stats on success
            stats.sent++;
            stats.received++;
            stats.min = Math.min(stats.min, latencyMs);
            stats.max = Math.max(stats.max, latencyMs);
            stats.sum += latencyMs;
            // record latency for percentile calculation
            if (config.percentile.length > 0) stats.records.push(latencyMs);

            resolve();
          })
          .on("error", function (e) {
            reject(e);
          })
          .on("timeout", function () {
            reject("timeout");
          })
          .connect(config.port, config.host);
      });
    } catch (e) {
      // update stats on failure
      stats.sent++;
      console.log(e);
    }

    await sleep(config.interval);
  }
}

export async function runHttp() {
  console.log(`HTTP mode, url: ${config.url}, count: ${config.count}`);
  console.log(); // empty line

  while (stats.sent < config.count) {
    try {
      await new Promise<void>((resolve, reject) => {
        const start = now();
        (config.url.startsWith("https") ? https : http)
          .request(config.url, { timeout: config.timeout }, (res) => {
            res.on("data", () => {}); // ignore response body. this function is required to receive response.
            res.on("close", () => {
              const latencyMs = (now() - start) / 1000;
              const result =
                (config.timestamp ? new Date().toISOString() + " " : "") +
                `get: ${latencyMs.toFixed(3)}ms, status: ${
                  res.statusCode
                }, length: ${res.headers["content-length"]}`;
              console.log(result);

              // update stats on success
              stats.sent++;
              stats.received++;
              stats.min = Math.min(stats.min, latencyMs);
              stats.max = Math.max(stats.max, latencyMs);
              stats.sum += latencyMs;
              // record latency for percentile calculation
              if (config.percentile.length > 0) stats.records.push(latencyMs);

              resolve();
            });
            res.on("error", (e) => reject(e));
          })
          .on("error", (e) => reject(e))
          .on("timeout", () => reject("timeout"))
          .end();
      });
    } catch (e) {
      // update stats on failure
      stats.sent++;
      console.log(e);
    }

    await sleep(config.interval);
  }
}


src/stats.ts
import { config } from "./config";

function getPercentile(sorted: readonly number[], percentile: number) {
  const index = Math.floor((percentile / 100) * sorted.length);
  return sorted[index];
}

export const stats = {
  sent: 0,
  received: 0,
  min: Infinity,
  max: 0,
  sum: 0,
  records: [] as number[],
};

export function statsToString() {
  // sort records for percentile calculation
  if (config.percentile.length > 0) stats.records.sort((a, b) => a - b);

  return (
    `statistics:\n` +
    [
      `sent: ${stats.sent}`,
      `received: ${stats.received}`,
      `success rate: ${((stats.received / stats.sent) * 100).toFixed(2)}%`,
    ].join(", ") +
    "\n" +
    [
      `min: ${stats.min.toFixed(3)}ms`,
      `max: ${stats.max.toFixed(3)}ms`,
      `avg: ${(stats.sum / stats.received).toFixed(3)}ms`,
      // percentile
      ...config.percentile.map(
        (p) => `p${p}: ${getPercentile(stats.records, p).toFixed(3)}ms`
      ),
    ].join(", ")
  );
}


src/config.ts
import yargs from "yargs";

const argv = yargs(process.argv.slice(2))
  .usage(
    "Usage: tcping [options] <host> [port]\nUsage (HTTP mode): tcping -h [options] <url>"
  )
  .options({
    i: { type: "number", alias: "interval", default: 500 },
    t: { type: "number", alias: "timeout", default: 3000 },
    p: {
      type: "number",
      alias: "port",
      default: 80,
      description: "Port number. Not effective in HTTP mode.",
    },
    c: {
      type: "number",
      alias: "count",
      default: -1,
      description: "Number of requests to send. -1 for infinite.",
    },
    P: {
      type: "string",
      alias: "percentile",
      default: "",
      description:
        'A list of percentile to display, separated by ",", e.g. "95,99".',
    },
    T: {
      type: "boolean",
      alias: "timestamp",
      default: false,
      description: "Display timestamp in output.",
    },
    h: {
      type: "boolean",
      alias: "http",
      default: false,
      description: "Use HTTP mode.",
    },
  })
  .parseSync();

export const config = Object.freeze({
  interval: argv.i,
  timeout: argv.t,
  port: (argv._[1] as number) ?? argv.p,
  host: argv._[0] as string,
  count: argv.c == -1 ? Infinity : argv.c,
  percentile: argv.P.split(",")
    .filter((x) => x.length > 0)
    .map((x) => Number(x)),
  timestamp: argv.T,
  http: argv.h,
  url: argv._[0] as string,
});
